File: pythonProject/neural_net.py
"""
Created on Mon Feb 22 15:04:12 2021

@author: Toby
@author: Pete Downey
"""
import threading

import numpy as np
import random
import copy
from keras import models
from keras import layers
from keras.models import model_from_json

from pythonProject import methods


class NeuralNet:
    nn_semaphore = threading.Semaphore()
    # code from:
    # https://machinelearningmastery.com/save-load-keras-deep-learning-models/
    json_file = open('data/model.json', 'r')
    assert json_file is not None
    "No NN model has been saved"
    loaded_model_json = json_file.read()
    json_file.close()
    _model = model_from_json(loaded_model_json)
    _model.load_weights("data/model.h5")


# TODO make sure this actually works
def get_model():
    NeuralNet.nn_semaphore.acquire()
    temp = copy.copy(NeuralNet._model)
    NeuralNet.nn_semaphore.release()
    return temp

def split_data(data, y_count):

    X = np.array(data[:, :, :-y_count + 1])
    Y = np.array(data[:, :, -y_count + 1:])

    return (X, Y)


def standerdize(x, y):
    x_new = copy.deepcopy(x)
    y_new = copy.deepcopy(y)

    for j, subD in enumerate(x_new):
        for i, n in enumerate(subD):
            mean = n.mean()
            ranges = n.max() - n.min()
            x_new[j, i] = ((n - mean) / ranges)
            y_new[j, i] = ((y_new[j, i] - mean) / ranges)
        # data2[j] = np.column_stack(
        #    (data2[j], np.array(range(len(data2[j])))))  # indices are kept track of to match each row to its inverse

    # TODO fix mean and range list
    return x_new, y_new


# randomly splits x and y from the third dimension
# EX (10, 20, 30) -> (10, 20, 11) (10, 20, 19)
def random_split_3rdD(x, y, splitpercent):
    # TODO make this better
    x1 = None
    x2 = None
    y1 = None
    y2 = None

    n1 = 0
    n2 = 0
    for n in range(0, x.shape[0]):

        if random.random() < splitpercent:
            if n1 == 0:
                x1 = [x[n, :, :]]
                y1 = [y[n, :, :]]
            else:
                x1 = np.append(x1, [x[n, :, :]], axis=0)
                y1 = np.append(y1, [y[n, :, :]], axis=0)
            n1 += 1
        else:
            if n2 == 0:
                x2 = [x[n, :, :]]
                y2 = [y[n, :, :]]
            else:
                x2 = np.append(x2, [x[n, :, :]], axis=0)
                y2 = np.append(y2, [y[n, :, :]], axis=0)
            n2 += 1

    return x1, x2, y1, y2


def full_standerdize(data, outputCount):
    # splits data into x and y
    x, y = split_data(data, outputCount + 1)

    # standardization
    x, y = standerdize(x, y)

    # splits x and y into the various types
    train_x, test_x, train_y, test_y = random_split_3rdD(x, y, 0.7)
    test_x, val_x, test_y, val_y = random_split_3rdD(test_x, test_y, 0.5)

    # i am using lambdas solely because it makes toby happy

    # reformat the data to work with the NN
    # adds an axis onto axis 1 creating a 4d array
    swap_and_expand = lambda data_input: np.expand_dims(data_input, axis=1)

    test_x = swap_and_expand(test_x)
    test_y = swap_and_expand(test_y)
    train_x = swap_and_expand(train_x)
    train_y = swap_and_expand(train_y)
    val_x = swap_and_expand(val_x)
    val_y = swap_and_expand(val_y)

    # converts the nn input format to its output format
    # gets rid of the volume data for y cause predictions only need to be x
    to_output_format = lambda data_input: np.swapaxes(np.squeeze(data_input[:, :, [0], :], axis=1), 1, 2)

    test_y = to_output_format(test_y)
    val_y = to_output_format(val_y)
    train_y = to_output_format(train_y)

    return test_x, test_y, train_x, train_y, val_x, val_y


# ASKTOBY
def train_neural_net(data):

    output_count = 50

    # splits up the data into the various types and seperates x and y and standerdizes
    test_x, test_y, train_x, train_y, val_x, val_y = full_standerdize(data, output_count)

    print("training")
    # honestly this is one of those things where the deeper you get the less it makes sense
    # the last output has an output of (node, outcount) ?????????????
    # don't even get me started on the convLSTM1D layer
    # MODEL ========
    model = models.Sequential()

    model.add(layers.ConvLSTM1D(50, kernel_size=20, strides=2,
                                input_shape=(1, 4, (1000 - output_count)),
                                activation='swish', return_sequences=False,
                                data_format='channels_first', padding="same"))

    model.add(layers.MaxPooling1D(pool_size=4, data_format='channels_first'))

    # hidden layers
    model.add(layers.Dense(128, activation='swish'))
    model.add(layers.Dense(64, activation='swish'))

    # output layer
    model.add(layers.Dense(1, activation='linear'))

    #model.summary()

    model.compile(loss='mean_squared_error', optimizer='adam', metrics=['accuracy'])

    # trains the model
    history = model.fit(train_x, train_y, epochs=30, batch_size=64, validation_data=(val_x, val_y), verbose=0)

    methods.plotLoss(history)

    # saves data
    # code from: https://machinelearningmastery.com/save-load-keras-deep-learning-models/
    model_json = model.to_json()
    with open("data/model.json", "w") as json_file:
        json_file.write(model_json)
    # serialize weights to HDF5
    model.save_weights("data/model.h5")

    NeuralNet.nn_semaphore.acquire()
    NeuralNet._model = model
    NeuralNet.nn_semaphore.release()


if __name__ == "__main__":
    train_neural_net()

File: model/neuralNet.cpp
//
// Created by Pete Downey on 9/15/21.
//
#include <iostream>
#include <thread>
#include <fstream>
#include <string>
#include <fstream>
#include <vector>
#include <utility> // std::pair
#include <stdexcept> // std::runtime_error
#include <sstream> // std::stringstream
#include <numeric>
#include <random>

#include <stdio.h>
#include <tensorflow/c/c_api.h>
#include <tensorflow/cc/ops/standard_ops.h>
//#include <tensorflow/core/framework/tensor.h>

using tensorflow::Scope;
//using tensorflow::Output;
//using tensorflow::Tensor;

//using tensorflow::ops::Const;
//using tensorflow::ops::MatMul;
//using tensorflow::ClientSession;

using namespace std;

vector<vector<float>> transpose(vector<vector<float>> data2) {

    vector<vector<float>> dataCopy( data2[0].size() , vector<float> (data2.size()));

    for (int y = 0; y < data2.size(); y++) {

        for (int x = 0; x < data2[0].size(); x++) {

            dataCopy[x][y] = data2[y][x];
        }
    }

    return dataCopy;
}

//modified  from
//https://www.gormanalysis.com/blog/reading-and-writing-csv-files-with-cpp/
vector<vector<float>> readCSV(std::string filename) {

    // Create an input filestream
    std::ifstream myFile(filename);

    // Make sure the file is open
    assert(myFile.is_open());
    assert(myFile.good());

    vector<vector<float>> data(130, vector<float>(1000, 0));

    for (int n = 0; n < 130; n++) {
        // Extract the lines of the file
        string singleLineData;
        std::getline(myFile, singleLineData);

        //why do you have to use a stringstrema
        stringstream singleLineDataStream(singleLineData);

        for (int n2 = 0; n2 < 1000; n2++) {
            string temp;
            std::string::size_type sz;

            std::getline(singleLineDataStream, temp, ',');
            data[n][n2] = stof(temp, &sz);
        }
    }

    // Close file
    myFile.close();

    return data;
}

std::tuple<std::vector<std::vector<float>>, std::vector<float>, std::vector<float>>
standerdize(vector<vector<float>> data) {

    vector<float> meanList = vector<float>();
    vector<float> rangeList = vector<float>();

    for (int n = 0; n < data.size(); n++) {

        float mean = std::reduce(data[0].begin(), data[0].end())/data[0].size();
        float ranges = *std::max_element(data[0].begin(), data[0].end()) -
                       *std::min_element(data[0].begin(), data[0].end());

        for (int n2 = 0; n2 < data.size(); n2++) {
            data[n][n2] = ((data[n][n2] - mean) / ranges);

            meanList.emplace_back(mean);
            rangeList.emplace_back(ranges);
        }
    }

    //TODO
    //data = np.column_stack(
    //        (data, np.array(range(len(data)))))  # indices are kept track of to match each row to its inverse

    return {data, meanList, rangeList};
}

std::tuple<std::vector<std::vector<float>>, std::vector<vector<float>>>
        splitData(std::vector<std::vector<float>> data, int splitPoint) {

    int xTill = (int)data[0].size() - splitPoint;

    std::vector<std::vector<float>> tData = transpose(data);

    std::vector<std::vector<float>> x(xTill, vector<float>(130));
    std::vector<std::vector<float>> y(1000 - xTill, vector<float>(130));


    for (int xData = 0; xData < data.size(); xData++) {

        if (xData < xTill)
            x[xData] = tData[xData];
        else
            y[xData - xTill] = tData[xData - xTill];
    }

    //FIXME: transpose spitting out error
    cout << x[130][50] << endl;

    x = transpose(x);
    y = transpose(y);
    return {x, y};
}

/**
 * splits data up into training and test data
 * @param xData xdata
 * @param yData y data
 * @param distribution how much is train vs test (0 - 100)
 * @return
 */
std::tuple<std::vector<std::vector<float>>, std::vector<vector<float>>, std::vector<vector<float>>, std::vector<vector<float>>>
splitDataRandom(std::vector<std::vector<float>> xData, std::vector<std::vector<float>> yData, int distribution) {

    assert(xData.size() == yData.size());

    std::vector<std::vector<float>> trainX, trainY, testX, testY;

    for (int n = 0; n < xData.size(); n++) {

        if (rand()%100 > distribution) {
            trainX.emplace_back(xData[n]);
            trainY.emplace_back(yData[n]);
        }
        else {
            testX.emplace_back(xData[n]);
            testY.emplace_back(yData[n]);
        }

    }

    return {trainX, testX, trainY, testY};
}


[[noreturn]] void NeuralNet() {

    // DATA ===========
    const char *fileName = "/Users/petedowney/Documents/GitHub/NNProject/data/data.csv";
    vector<vector<float>> data = readCSV(fileName);


    // standardization
    auto [data2, meanList, rangeList] = standerdize(data);

    //I hate this
    data = data2;
    data2.clear();

    auto [xData, yData] = splitData(data, 50 + 1);

    auto [trainXTemp, testX, trainYTemp, testY] = splitDataRandom(xData, yData, 30);
    auto [trainX, valX, trainY, valY] = splitDataRandom(trainXTemp, trainYTemp, 50);

    trainXTemp.clear();
    trainYTemp.clear();


    auto root = Scope::NewRootScope();


    /*

// split data

    reshaped = lambda;

    x.reshape(x.shape[0], x.shape[1], 1);

    test_X = reshaped(test_X);
    train_X = reshaped(train_X);
    val_X = reshaped(val_X);

    test_Y, test_inverse = methods.snipY(test_Y);
    train_Y, train_inverse = methods.snipY(train_Y)
    val_Y, val_inverse = methods.snipY(val_Y);

// MODEL ========
    model = models.Sequential();

// input layer (pre-network convolution)
    model.add(layers.Conv1D(32, kernel_size = 8, strides = 1, input_shape = (None, 1), activation = 'swish',
                            padding = "causal"));
    model.add(layers.AveragePooling1D(2));

// LSTM
    model.add(layers.LSTM(48, activation = 'swish', input_shape = (None, 1), return_sequences = False));

// hidden layers
    model.add(layers.Dense(128, activation = 'swish'));
    model.add(layers.Dense(64, activation = 'linear'));

// output layer
    model.add(layers.Dense(output_count, activation = 'linear'));


    model.compile(loss = 'mean_squared_error', optimizer = 'adam', metrics =['accuracy']);

    history = model.fit(train_X, train_Y, epochs = 30, batch_size = 64, validation_data = (val_X, val_Y), verbose = 0);
*/

}

